### PR TITLE
Remote content endpoint streaming

### LINKED ISSUE #9386: Create a /content/storage endpoint that will proxy to a remote server
An API endpoint setup under `/remote/content/storage` that will proxy to a specific remote server using the get param `baseurl` specified. If no `baseurl` is specified, it should return a 404.

The API endpoint will then make a further request to `<baseurl>/content/storage/....` and then stream the response from that request as the response to the request being made from the client.

To intelligently cache these resources, this endpoint should first defer to Kolibri's whitenoise implementation to serve up the file - if that returns a 404, then it should attempt to make the request described above.

If the request is successful, and disk space is available and writable, the resource should be cached in the local Kolibri content storage.

This is probably best implemented within the core Kolibri whitenoise implementation, so that it can be used in both the main wsgi and alternate origin wsgi servers.

### LINKED ISSUE #9388: Handle range requests on remote content serving
To better support video streaming from remote servers, we should support proxying range requests, and caching the results of those range requests (where eventually with enough range requests, we'll have a whole file!).

### PR DESCRIPTION
**Pull request by rtibbles**
## Summary
* Moves the file transfer utilities into the `kolibri.utils` namespace so that they can be imported without importing the Django stack
* Makes some updates to the file transfer utilities to allow them to be used solely via `next` and not just via explicit iteration in a loop.
* Catches errors that are thrown when a file transfer attempts to move to its destination, but the source transfer file has disappeared, if the destination file already exists, it silently swallows the error - this allows for multiple simultaneous transfers of the same file to co-exist and not stomp each other.
* Adds additional functionality to the DynamicWhitenoise class, whereby it will accept a `baseurl` GET parameter, which, if the file does not exist on disk, will attempt to fetch the file using the `FileDownload` file transfer class.
* As the file is downloading, the file download is wrapped in a special filelike wrapper object that streams the download into the response.
* This wrapper object supports `seek` in order to support range requests, but still downloads the entire file, as it seems complex to construct the full file out of multiple potentially non-sequential range requests.

Note: Cleaning up the downloaded files is deferred to a follow up issue!

## References
Fixes #9386
Fixes #9388

## Reviewer guidance
Test this out by streaming some files directly from Studio, e.g.:
http://127.0.0.1:8000/content/storage/1/1/11e1c2c83b9dd46e584d75f484386021.mp4?baseurl=https://studio.learningequality.org
http://127.0.0.1:8000/content/storage/9/d/9d523f8a736e372fbe10ac895085fce9.png?baseurl=https://studio.learningequality.org
http://127.0.0.1:8000/content/storage/8/a/8a9c9011e527c0ebf9b09d642eba5f78.pdf?baseurl=https://studio.learningequality.org

Confirm that reloading the file also works.
Delete the file from storage and try the request again, confirm that it works.

Question: Is the range request implementation good enough?

----

## Testing checklist

- [x] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [x] Critical and brittle code paths are covered by unit tests

## PR process

- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

File: kolibri/utils/file_transfer.py [hunk 1]
```diff
@@ -36,7 +72,7 @@ def __init__(
         block_size=2097152,
         remove_existing_temp_file=True,
         timeout=DEFAULT_TIMEOUT,
-        cancel_check=None,
+        cancel_check=lambda: False,
```
Comment (jredrejo) [inline kolibri/utils/file_transfer.py]: what's the advantage of this change in the default value? I only see cancel_check being used in `if` sentences, so None works fine there too

File: kolibri/utils/kolibri_whitenoise.py [hunk 2]
```diff
@@ -182,6 +296,12 @@ def find_and_cache_dynamic_file(self, url):
                     except (IOError, OSError):
                         pass
                 self.add_file_to_dictionary(url, path, stat_cache=stat_cache)
+        elif (
+            remote_baseurl is not None
+            and self.writable_check is not None
+            and self.writable_check.match(url)
+        ):
+            self.files[url] = self.get_streaming_static_file(url, remote_baseurl)
```
Comment (jredrejo) [inline kolibri/utils/kolibri_whitenoise.py:304]: this is really nice

File: kolibri/utils/file_transfer.py [hunk 3]
```diff
@@ -107,7 +157,11 @@ def next(self):
         return chunk
 
     def _move_tmp_to_dest(self):
-        shutil.move(self.dest_tmp, self.dest)
+        try:
+            shutil.move(self.dest_tmp, self.dest)
+        except (IOError, OSError) as e:
```
Comment (jredrejo) [inline kolibri/utils/file_transfer.py]: While we keep using Python 2.7 I like to add at the beginning of the modules having code opening files, this:
```
try:
    FileNotFoundError
except NameError:
    FileNotFoundError = IOError
```
and use `FileNotFound` as a different error catching, what can be more informative in the places where  you're using `IOError` when python 3 is used

**Review by jredrejo — APPROVED** [review]
Code looks good to me, and I haven't found any problem while testing it.
I've added a couple of comments, but  none of them are a blocker, they can be used or discarded without any problem.

File: kolibri/utils/file_transfer.py (hunk 1 above)
Comment (rtibbles) [reply kolibri/utils/file_transfer.py]: I could have changed it to be `if cancel_check and cancel_check()` where it is used - I can update it to that, I was just being slightly lazy.

File: kolibri/utils/file_transfer.py (hunk 3 above)
Comment (rtibbles) [reply kolibri/utils/file_transfer.py]: Yeah, that's a nice practice, will update.

**Pull request merged**